docs/proposals/scheduling/20231205-dynamic-reservation-provisioning.md

```diff
@@ -0,0 +1,181 @@
+
+### Non-Goals/Future Work
+
+1. None
+
+## User stories
+
+### Use Case 1: Just-in-Time Resource Allocation for High-Priority Jobs
+
+Scenario: A high-priority job needs to be scheduled immediately, but there are no pre-created Reservation CRDs that can guarantee the resources it needs.
+
+Outcome with DRP: As the scheduler attempts to place the Pod for this job, DRP triggers the creation of a Reservation CRD tailored to the job's resource requirements. The job is scheduled without delay, ensuring that critical workloads receive the resources they need without the need for pre-planning.
+
+### Use Case 2: Efficient Resource Utilization for Fluctuating Workloads
+
+Scenario: Workloads in a cluster exhibit significant variance in resource demands, making it difficult to predict and pre-create suitable Reservation CRDs.
+
+Outcome with DRP: DRP allows the system to adapt to fluctuating workloads by creating Reservations dynamically. This results in better resource utilization and avoids the overhead of managing a multitude of static reservations that may not align with actual usage.
+
+### Use Case 3: Simplified Operations for DevOps Teams
+
+Scenario: DevOps teams need to deploy applications quickly without being bottlenecked by reservation management.
+
+Outcome with DRP: Teams can deploy Pods with annotations indicating their resource requirements, and DRP ensures the necessary Reservation CRDs are created in lockstep with Pod scheduling. This reduces the operational complexity for DevOps teams and accelerates the deployment pipeline.
+
+### Use Case 4: On-Demand Scaling for Microservices
+
+Scenario: A microservice architecture requires certain services to scale up rapidly in response to spikes in demand, which may not have corresponding Reservation CRDs in place.
+
+Outcome with DRP: DRP facilitates on-demand scaling by provisioning reservations as microservices scale out. This ensures that each service instance has the resources it needs to perform optimally without pre-allocating resources.
+
+
+## Proposal
+
+### Overview
+
+The DRP mechanism will work as follows:
+
+1. Pod Annotation for Dynamic Reservation: Users can annotate Pods with `scheduling.koordinator.sh/reservation-template`, which includes the specification for a potential Reservation CRD.
+2. Scheduler Responsiveness: If an incoming Pod has this annotation and there is no available Reservation CRD instance that fits its requirements, the Koordinator scheduler will construct a virtual reservation object in memory. This object will emulate the behavior of a Reservation CRD instance to temporarily reserve resources for the Pod.
+3. Triggering Reservation Creation: In the PreBind phase of scheduling, the scheduler appends an annotation `scheduling.koordinator.sh/new-reservation-requirement` to the Pod, signaling an external provisioner to create an actual Reservation CRD instance based on the provided template.
```

Comment:
Why doesn't the scheduler create this reservation directly?

Reply:
When I re-reviewed this process yesterday, I also considered whether it should be created directly by the scheduler? Discussing with @FillZpp  , the external provisioner may need to complete more complex custom logic in the Reservation creation process.

Reply:
The provisioner(controller) may have its specific protocol on the Reservations it owned, e.g., labels, annotations, and also the Reservation might have to be owned by another owner reference.

Scheduler should not care about these things. It only helps to reserve resources, and let those provisioners create the Reservations they want.

Reply:
The provisioner(controller) may have its specific protocol on the Reservations it owned, e.g., labels, annotations, and also the Reservation might have to be owned by another owner reference.

Scheduler should not care about these things. It only helps to reserve resources, and let those provisioners create the Reservations they want.

Reply:
When creating externally, there are a series of questions that need to be answered, such as:
- If there is no response externally, what should be done? Rollback and how long will it take to do it?
- What to do if the reservation created externally is inconsistent with the resources occupied by the scheduler?

Reply:
Such exceptions are indeed possible, and should they occur, we can consider adopting the following fault tolerance mechanisms:

1. Expose any exceptions in Reservation creation through Events or Metrics.
2. The scheduler should have a default wait timeout, similar to the mechanism in Dynamic Volume Provisioning. If a Reservation is not created after waiting for 10 minutes, the process should be rolled back, and the Pod should be re-queued for another scheduling attempt.
3. Consider retaining the virtual reservation object; it could be reused in the next Pod scheduling attempt. It is recommended to keep it at first. Once the provisioner recovers from any exceptions, it can attempt to create a new Reservation object that can replace the virtual reservation object in memory.
4. In terms of the resource delivery success rate, if the creation of a Reservation fails, the Pod's scheduling could still be completed, and the virtual reservation object should be deleted.